### ISSUE TITLE
Camunda8 - Helm Upgrade Fails

### ISSUE THREAD

**Issue by TortoraLuigi** [camunda/camunda-platform-helm#309]
helm upgrade fails for 3 Times because of missing parameters and after that it fails completly.

**_helm upgrade camunda8 camunda8/camunda-platform --values values8.yaml_**

_Error: UPGRADE FAILED: execution error at (camunda-platform/charts/identity/templates/tasklist-secret.yaml:10:22): 
PASSWORDS ERROR: You must provide your current passwords when upgrading the release.
                 Note that even after reinstallation, old credentials may be needed as they may be kept in persistent volume claims.
                 Further information can be obtained at https://docs.bitnami.com/general/how-to/troubleshoot-helm-chart-issues/#credential-errors-while-upgrading-chart-releases

    'global.identity.auth.tasklist.existingSecret' must not be empty, please add '--set global.identity.auth.tasklist.existingSecret=$TASKLIST_SECRET' to the command. To get the current value:

        export TASKLIST_SECRET=$(kubectl get secret --namespace "default" "camunda8-tasklist-identity-secret" -o jsonpath="{.data.tasklist-secret}" | base64 --decode)_

_**export TASKLIST_SECRET=$(kubectl get secret --namespace "default" "camunda8-tasklist-identity-secret" -o jsonpath="{.data.tasklist-secret}" | base64 --decode)
helm upgrade camunda8 camunda8/camunda-platform --values values8.yaml --set global.identity.auth.tasklist.existingSecret=$TASKLIST_SECRET**_

_Error: UPGRADE FAILED: execution error at (camunda-platform/charts/identity/templates/optimize-secret.yaml:10:22): 
PASSWORDS ERROR: You must provide your current passwords when upgrading the release.
                 Note that even after reinstallation, old credentials may be needed as they may be kept in persistent volume claims.
                 Further information can be obtained at https://docs.bitnami.com/general/how-to/troubleshoot-helm-chart-issues/#credential-errors-while-upgrading-chart-releases

    'global.identity.auth.optimize.existingSecret' must not be empty, please add '--set global.identity.auth.optimize.existingSecret=$OPTIMIZE_SECRET' to the command. To get the current value:

        export OPTIMIZE_SECRET=$(kubectl get secret --namespace "default" "camunda8-optimize-identity-secret" -o jsonpath="{.data.optimize-secret}" | base64 --decode)_

_**export OPTIMIZE_SECRET=$(kubectl get secret --namespace "default" "camunda8-optimize-identity-secret" -o jsonpath="{.data.optimize-secret}" | base64 --decode)
helm upgrade camunda8 camunda8/camunda-platform --values values8.yaml --set global.identity.auth.tasklist.existingSecret=$TASKLIST_SECRET --set global.identity.auth.optimize.existingSecret=$OPTIMIZE_SECRET**_

Error: UPGRADE FAILED: execution error at (camunda-platform/charts/identity/templates/operate-secret.yaml:10:21): 
PASSWORDS ERROR: You must provide your current passwords when upgrading the release.
                 Note that even after reinstallation, old credentials may be needed as they may be kept in persistent volume claims.
                 Further information can be obtained at https://docs.bitnami.com/general/how-to/troubleshoot-helm-chart-issues/#credential-errors-while-upgrading-chart-releases

    'global.identity.auth.operate.existingSecret' must not be empty, please add '--set global.identity.auth.operate.existingSecret=$OPERATE_SECRET' to the command. To get the current value:

_**export OPERATE_SECRET=$(kubectl get secret --namespace "default" "camunda8-operate-identity-secret" -o jsonpath="{.data.operate-secret}" | base64 --decode)
helm upgrade camunda8 camunda8/camunda-platform --values values8.yaml --set global.identity.auth.tasklist.existingSecret=$TASKLIST_SECRET --set global.identity.auth.optimize.existingSecret=$OPTIMIZE_SECRET --set global.identity.auth.operate.existingSecret=$OPERATE_SECRET**_

_Error: UPGRADE FAILED: cannot patch "camunda8-identity" with kind Deployment: Deployment.apps "camunda8-identity" is invalid: [spec.template.spec.containers[0].env[4].valueFrom.secretKeyRef.name: Invalid value: "91PjdWKZ5p": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'), spec.template.spec.containers[0].env[7].valueFrom.secretKeyRef.name: Invalid value: "HFuYimnFrg": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'), spec.template.spec.containers[0].env[10].valueFrom.secretKeyRef.name: Invalid value: "7niyS1tPNi": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')] && cannot patch "camunda8-operate" with kind Deployment: Deployment.apps "camunda8-operate" is invalid: spec.template.spec.containers[0].env[4].valueFrom.secretKeyRef.name: Invalid value: "91PjdWKZ5p": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*') && cannot patch "camunda8-optimize" with kind Deployment: Deployment.apps "camunda8-optimize" is invalid: spec.template.spec.containers[0].env[8].valueFrom.secretKeyRef.name: Invalid value: "7niyS1tPNi": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*') && cannot patch "camunda8-tasklist" with kind Deployment: Deployment.apps "camunda8-tasklist" is invalid: spec.template.spec.containers[0].env[4].valueFrom.secretKeyRef.name: Invalid value: "HFuYimnFrg": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*') && cannot patch "camunda8-tasklist" with kind Ingress: admission webhook "validate.nginx.ingress.kubernetes.io" denied the request: host "_" and path "/" is already defined in ingress default/camunda8-operate_

**Comment by Zelldon** [comment]
Hey @TortoraLuigi 

thanks for reporting! Sorry for the issues you're facing, did you follow the guide https://docs.bitnami.com/general/how-to/troubleshoot-helm-chart-issues/#credential-errors-while-upgrading-chart-releases here? You need to rollback, retrieve your previous generated secrets, and set them in upgrade otherwise your apps will no longer be accessible. The issue is that on the first installation passwords are generated if not specified and they are persisted on the PVCs. Helm has no better way to solve this in a good way. We use the common library from bitnami to generate and access secrets, which show you the error message you're seeing.

**Comment by AndersRunningen** [comment]
@Zelldon what if the secrets where configured with a lookup function like this instead?
<https://itnext.io/manage-auto-generated-secrets-in-your-helm-charts-5aee48ba6918>

Pseudo code would be something like:
1. Check if existingSecret is used, if it is, use that
2. Is the autogenerated secret with the default name in place? no, create and use it, yes, get the secret object and use it.

**Comment by Zelldon** [comment]
Thanks for sharing @AndersRunningen 

We use the commons lib chart from bitnami https://github.com/bitnami/charts/tree/master/bitnami/common which already does an lookup for existing this https://github.com/bitnami/charts/blob/master/bitnami/common/templates/_secrets.tpl#L93

You can always bring your own secrets btw see here https://github.com/camunda/camunda-platform-helm/blob/main/charts/camunda-platform/values.yaml#L79 and then no secret will be generated.

**Comment by Zelldon** [comment]
Just do document some of my findings and potential upcoming fixes. 

I started to write an upgrade guide for the Helm charts, you can see the draft here https://github.com/camunda/camunda-platform-docs/pull/880


After following the instructions and setting all necessary secrets I run into the issue which was reported here:

```sh
$ helm upgrade camunda-platform-test charts/camunda-platform/ --set global.identity.auth.tasklist.existingSecret=$TASKLIST_SECRET --set global.identity.auth.optimize.existingSecret=$OPTIMIZE_SECRET --set global.identity.auth.operate.existingSecret=$OPERATE_SECRET
Error: UPGRADE FAILED: cannot patch "camunda-platform-test-identity" with kind Deployment: Deployment.apps "camunda-platform-test-identity" is invalid: spec.template.spec.containers[0].env[4].valueFrom.secretKeyRef.name: Invalid value: "dIqbP5OmYH": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
```

The problem is that our current configuration of the `existingSecret` doesn't work with the instructions given by Bitnami, which shows the following:

```
    'global.identity.auth.tasklist.existingSecret' must not be empty, please add '--set global.identity.auth.tasklist.existingSecret=$TASKLIST_SECRET' to the command. To get the current value:
        export TASKLIST_SECRET=$(kubectl get secret --namespace "zell-c8-optimize" "camunda-platform-test-tasklist-identity-secret" -o jsonpath="{.data.tasklist-secret}" | base64 --decode)
```

We expect that the `existingSecret` references a secret via name, see here https://github.com/camunda/camunda-platform-helm/blob/main/charts/camunda-platform/charts/identity/templates/deployment.yaml#L43 and here the related definition https://github.com/camunda/camunda-platform-helm/blob/main/charts/camunda-platform/values.yaml#L77-L79

What we actually want on the upgrade is that the value from `existingSecret` is used to recreate the secret with the same value. This is actually can be done via the template `common.secrets.passwords.manage` we use here https://github.com/camunda/camunda-platform-helm/blob/main/charts/camunda-platform/charts/identity/templates/operate-secret.yaml#L10. Funny is that we have excluded this secret if the `existingSecret` is set, see here https://github.com/camunda/camunda-platform-helm/blob/main/charts/camunda-platform/charts/identity/templates/operate-secret.yaml#L1


If you want to take a look at the implementation of the template check this https://github.com/bitnami/charts/blob/master/bitnami/common/templates/_secrets.tpl#L85
Usage of this in other charts you can see here in Keycloak https://github.com/bitnami/charts/blob/master/bitnami/keycloak/templates/secrets.yaml#L18 or PostgreSQL https://github.com/bitnami/charts/blob/master/bitnami/postgresql/templates/secrets.yaml#L17

I was wondering why we see only the issues on our side, not on keycloak nor PostgreSQL. I think this is related to the Helm upgrade mechanism. 

https://helm.sh/docs/intro/using_helm/
> It will only update things that have changed since the last release.

I guess due to several labels we use, including versions etc. helm will recreate these secrets on upgrade, which cause the exception. If we would add other configs or bump the keycloak version we might see this as well on Keycloak. Keycloak and Postgresql suggest anyway to store the secrets and provide them on upgrade. 

You can read that in the following resources:

 * https://docs.bitnami.com/general/how-to/troubleshoot-helm-chart-issues/#credential-errors-while-upgrading-chart-releases
 * https://docs.bitnami.com/kubernetes/infrastructure/postgresql/administration/upgrade/

----


### Proposal

In order to solve this issue, I have the following proposal:

If the `existingSecret` contains a string we use it as value for our secret creation, for that we can check on some places for the type `string`. In order to still support bring your own secrets, we expect the following format:

```
existingSecret:
  name: <secretname>
```

This allows us to distinguish between both cases and still reference user secrets.

#### Example

Example change on the identity deployment for referencing the operate secret:

```diff
          - name: KEYCLOAK_INIT_OPERATE_SECRET
-           {{- if .Values.global.identity.auth.operate.existingSecret }}
+          {{- if and .Values.global.identity.auth.operate.existingSecret (not (typeIs "string" .Values.global.identity.auth.operate.existingSecret)) }}
            valueFrom:
              secretKeyRef:
                {{- /*
                    Helper: https://github.com/bitnami/charts/blob/master/bitnami/common/templates/_secrets.tpl
                    Usage in keycloak secrets https://github.com/bitnami/charts/blob/master/bitnami/keycloak/templates/secrets.yaml
                    and in statefulset https://github.com/bitnami/charts/blob/master/bitnami/keycloak/templates/statefulset.yaml
                */}}
                name: {{ include "common.secrets.name" (dict "existingSecret" .Values.global.identity.auth.operate.existingSecret "context" $) }}
                key: operate-secret
            {{- else }}
            valueFrom:
              secretKeyRef:
                name: {{ include "identity.secretNameOperateIdentity" . }}
                key: operate-secret
            {{- end }}
```

If `.Values.global.identity.auth.operate.existingSecret` is defined and not a string then we will use the upper path and use the `existingSecret` as reference otherwise we know we created the secret manifest and the value will be used their. 

In order to deploy the secret in such case we have to change the condition here as well:

```diff
+{{- if and (.Values.global.identity.auth.enabled) (or (not .Values.global.identity.auth.operate.existingSecret) (typeIs "string" .Values.global.identity.auth.operate.existingSecret)) }}
-{{- if and (.Values.global.identity.auth.enabled) (not .Values.global.identity.auth.operate.existingSecret) }}
{{- $secretName := include "identity.secretNameOperateIdentity" . }}
apiVersion: v1
kind: Secret
metadata:
  name: {{ $secretName }}
  labels: {{- include "identity.labels" . | nindent 4 }}
type: Opaque
data:
  operate-secret: {{ include "common.secrets.passwords.manage" (dict "secret" $secretName "key" "operate-secret" "length" 10 "providedValues" (list "global.identity.auth.operate.existingSecret") "context" $) }}
{{- end }}
```

We have to do this of course for all components and add some new tests.


@oleschoenburg I just documented here my thoughts and potential solution lets discuss this later.

**Comment by Zelldon** [comment]
Thanks for reporting @TortoraLuigi with the recent fix #311 it should work now. I created an integration tests which does an upgrade you can see that here https://github.com/camunda/camunda-platform-helm/blob/main/charts/camunda-platform/test/integration/integration_test.go#L111-L149

To make sure that the upgrade works as expected make sure to set all secrets as you can see here

https://github.com/camunda/camunda-platform-helm/blob/main/charts/camunda-platform/test/integration/integration_test.go#L130-L135

I'm working on an upgrade guide in the docs which will described this in more detail
